crates/webidl/src/first_pass.rs

```diff
@@ -103,43 +114,177 @@ impl FirstPass for webidl::ast::Enum {
     }
 }
 
-impl FirstPass for webidl::ast::Interface {
-    fn first_pass<'a>(&'a self, record: &mut FirstPassRecord<'a>) -> Result<()> {
+impl FirstPass<()> for webidl::ast::Interface {
+    fn first_pass<'a>(&'a self, record: &mut FirstPassRecord<'a>, (): ()) -> Result<()> {
         use webidl::ast::Interface::*;
 
         match self {
-            NonPartial(interface) => interface.first_pass(record),
-            _ => {
-                // Other interfaces aren't currently used in the first pass
+            Partial(interface) => interface.first_pass(record, ()),
+            NonPartial(interface) => interface.first_pass(record, ()),
+            // TODO
+            Callback(..) => {
+                warn!("Unsupported WebIDL interface: {:?}", self);
                 Ok(())
             }
         }
     }
 }
 
-impl FirstPass for webidl::ast::NonPartialInterface {
-    fn first_pass<'a>(&'a self, record: &mut FirstPassRecord<'a>) -> Result<()> {
-        if record.interfaces.insert(self.name.clone()) {
-            warn!("Encountered multiple declarations of {}", self.name);
+impl FirstPass<()> for webidl::ast::NonPartialInterface {
+    fn first_pass<'a>(&'a self, record: &mut FirstPassRecord<'a>, (): ()) -> Result<()> {
+        if record.interfaces.contains_key(&self.name) {
```

Comment:
Could this perhaps use `record.interfaces.entry(...).or_insert_with(...)` to deduplicate these branches a bit?

Reply:
I wanted to do something like https://play.rust-lang.org/?gist=3c1e682a8747dd4c2a70c8b9f81898fc, but this does not compile.